### PR TITLE
[FEATURE] Create profiler from DataContext

### PR DESCRIPTION
**Pull request by cdkini**
Changes proposed in this pull request:
- `add_profiler` method and related tests

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/data_context/data_context.py [hunk 1]
```diff
@@ -3259,6 +3263,32 @@ def run_checkpoint(
...
+                "rules": rules,
+                "class_name": class_name,
+                "module_name": module_name,
+                "variables": variables,
+                "commented_map": commented_map,
+            }
+        )
+        return profiler_toolkit.add_profiler(
+            config=config, data_context=self, ge_cloud_id=ge_cloud_id
+        )
```
Comment (cdkini) [inline great_expectations/data_context/data_context.py]: Should be `RuleBasedProfiler.add_profiler` class method but putting everything in toolkit and saving all the updates for a refactoring PR.

File: great_expectations/data_context/data_context.py [hunk 2]
```diff
@@ -3264,6 +3270,41 @@ def run_checkpoint(
...
+        }
+        # Package config and perform validation
+        errors: Dict[str, List[str]] = ruleBasedProfilerConfigSchema.validate(
+            config_data
+        )
+        if errors:
+            raise ValidationError(
+                f"Something went wrong during profiler validation: {errors}"
+            )
+        config: RuleBasedProfilerConfig = RuleBasedProfilerConfig(**config_data)
```
Comment (cdkini) [inline great_expectations/data_context/data_context.py]: This part is up to debate but I think packaging a config object and having the RBP util method work that single object is much cleaner than accepting each single argument.

@alexsherstinsky let's also discuss `validate` and what makes sense here.

This method requires a test - this PR is to be blocked until that path is under test.

File: great_expectations/rule_based_profiler/toolkit.py [hunk 3]
```diff
@@ -1,19 +1,82 @@
...
+            configuration_key=config.name,
+        )
+
+    profiler_store = data_context.profiler_store
+    profiler_ref = profiler_store.set(key=key, value=config)
+    if isinstance(profiler_ref, GeCloudIdAwareRef):
+        pass
+        # Chetan - 20220201 - Open to determine how to treat ge_cloud_id
+        # ge_cloud_id = profiler_ref.ge_cloud_id
+        # new_profiler.ge_cloud_id = uuid.UUID(ge_cloud_id)
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/toolkit.py]: ge_cloud_id is a tricky matter so commenting out and requesting a conversation to determine the relationship between GE Cloud and RBP

File: great_expectations/rule_based_profiler/toolkit.py [hunk 4]
```diff
@@ -1,19 +1,82 @@
...
+
+    return new_profiler
+
+
+def _check_validity_of_batch_requests_in_config(
+    config: RuleBasedProfilerConfig, profiler_store_name: str
+) -> None:
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/toolkit.py]: Util to perform the same check that Checkpoint does (but in an RBP-specific way)

File: tests/conftest.py [hunk 5]
```diff
@@ -5046,6 +5047,16 @@ def profiler_key(profiler_name: str) -> ConfigurationIdentifier:
...
+@pytest.fixture(scope="function")
+def ge_cloud_profiler_id() -> str:
+    return "my_ge_cloud_profiler_id"
+
+
+@pytest.fixture(scope="function")
+def ge_cloud_profiler_key(ge_cloud_profiler_id: str) -> GeCloudIdentifier:
+    return GeCloudIdentifier(resource_type="contract", ge_cloud_id=ge_cloud_profiler_id)
+
+
```
Comment (cdkini) [inline tests/conftest.py]: Necessary to test the GE Cloud specific code paths.

File: great_expectations/rule_based_profiler/toolkit.py [hunk 6]
```diff
@@ -1,19 +1,82 @@
...
 from great_expectations.data_context.util import instantiate_class_from_config
 from great_expectations.rule_based_profiler import RuleBasedProfiler
-from great_expectations.rule_based_profiler.config import RuleBasedProfilerConfig
+from great_expectations.rule_based_profiler.config.base import RuleBasedProfilerConfig
 from great_expectations.util import filter_properties_dict
 
 
+def add_profiler(
+    config: RuleBasedProfilerConfig,
+    data_context: "DataContext",  # noqa: F821
```
Comment (donaldheppner) [inline great_expectations/rule_based_profiler/toolkit.py:20]: Can we remove DataContext from this method and instead pass the dependencies needed to add_profiler (like the cloud_mode, profiler store).

File: great_expectations/rule_based_profiler/toolkit.py [hunk 7]
```diff
@@ -1,19 +1,82 @@
...
+            batch_requests.append(domain_builder["batch_request"])
+
+        parameter_builders = rule.get("parameter_builders", [])
+        for parameter_builder in parameter_builders:
+            if "batch_request" in parameter_builder:
+                batch_requests.append(parameter_builder["batch_request"])
+
+    # DataFrames shouldn't be saved to ProfilerStore
+    for batch_request in batch_requests:
+        if batch_request_contains_batch_data(batch_request=batch_request):
```
Comment (donaldheppner) [inline great_expectations/rule_based_profiler/toolkit.py:72]: "check" methods shouldn't throw exceptions if the check fails; they should return a bool, potentially a bool,str tuple

**Review by donaldheppner — CHANGES_REQUESTED** [review]

File: great_expectations/rule_based_profiler/toolkit.py (hunk 7 above)
Comment (cdkini) [reply great_expectations/rule_based_profiler/toolkit.py:72]: I agree, great call! I've stuck to a simple boolean but let me know if you want a more verbose error message.

File: great_expectations/rule_based_profiler/toolkit.py (hunk 6 above)
Comment (cdkini) [reply great_expectations/rule_based_profiler/toolkit.py:20]: I was aiming to do that but the RBP constructor takes in a DataContext so I've opted to pass it in through `instantiate_class_from_config`. I don't love the design though.

@alexsherstinsky can I get away with omitting the `runtime_environment` portion of this:
```python
    profiler = instantiate_class_from_config(
        config=config,
        runtime_environment={
            "data_context": data_context,
        },
        config_defaults={
            "module_name": "great_expectations.rule_based_profiler",
        },
    )

```

File: great_expectations/data_context/data_context.py [hunk 8]
```diff
@@ -20,6 +20,8 @@
 from ruamel.yaml.comments import CommentedMap
 from ruamel.yaml.constructor import DuplicateKeyError
 
+from great_expectations.marshmallow__shade import ValidationError
```
Comment (alexsherstinsky) [inline great_expectations/data_context/data_context.py]: @cdkini It seems to be a repeat of the import below; am I missing something?  Thanks!

File: great_expectations/data_context/data_context.py [hunk 9]
```diff
@@ -120,6 +122,10 @@
 from great_expectations.render.renderer.site_builder import SiteBuilder
 from great_expectations.rule_based_profiler import RuleBasedProfiler
 from great_expectations.rule_based_profiler.config import RuleBasedProfilerConfig
+from great_expectations.rule_based_profiler.config.base import (
```
Comment (alexsherstinsky) [inline great_expectations/data_context/data_context.py]: @cdkini Per #4111 this will change the location (which is partly why we want to merge that PR first).  Thanks!

File: great_expectations/data_context/data_context.py [hunk 10]
```diff
@@ -3264,6 +3270,41 @@ def run_checkpoint(
...
+    def add_profiler(
+        self,
+        name: str,
+        config_version: float,
+        rules: Dict[str, dict],
+        class_name: str,
+        module_name: Optional[str] = None,
+        variables: Optional[dict] = None,
+        commented_map: Optional[CommentedMap] = None,
+        ge_cloud_id: Optional[str] = None,
```
Comment (alexsherstinsky) [inline great_expectations/data_context/data_context.py:3279]: @cdkini Are we going ahead with adding `ge_cloud_id`?  If so, then please do so consistently across all the relevant classes, just like in `Checkpoint` (after #4111 is merged, it will be easy to touch-up all involved code sections).  However, we definitely need to come up with a cleaner design.  Thanks!  /cc @donaldheppner

File: great_expectations/data_context/data_context.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/data_context/data_context.py]: @cdkini My preference would be for now to stick to the battle-tested "deserialize/serialize" (load/dump) pattern.  We are going to do a lot of rearchitecting of configuration and serialization, and I would feel more comfortable if we deferred this validation to that time as well (it will allow us to learn more about the applicability of each; moreover, it will be a great opportunity to consider upgrading to the latest "shipping" Marshmallow, instead of relying on our own patched rendition of an older version -- the latest release may give us even better validation tools).

File: great_expectations/data_context/data_context.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/data_context/data_context.py]: @cdkini I think your approach is perfect for today's needs.

File: great_expectations/data_context/data_context.py [hunk 11]
```diff
@@ -3264,6 +3270,41 @@ def run_checkpoint(
...
+    ):
+        config_data = {
```
Comment (alexsherstinsky) [inline great_expectations/data_context/data_context.py:3281]: @cdkini Mind adding the type hint (for aesthetics purposes, of course).  Thanks! :-)

File: great_expectations/rule_based_profiler/toolkit.py (hunk 6 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/toolkit.py:20]: @cdkini I believe that -- indeed -- `runtime_environment` is exactly the correct way to pass in the `data_context` -- it was intended for such purposes; so it looks like you got it!

File: great_expectations/rule_based_profiler/toolkit.py [hunk 12]
```diff
@@ -2,18 +2,80 @@
...
+    ge_cloud_id: Optional[str] = None,
+) -> RuleBasedProfiler:
+    if not _check_validity_of_batch_requests_in_config(config):
+        raise ge_exceptions.InvalidConfigError(
+            f'batch_data found in batch_request cannot be saved to ProfilerStore "{data_context.profiler_store_name}"'
+        )
+
+    profiler_config: dict = config.to_json_dict()
+
+    new_profiler = instantiate_class_from_config(
```
Comment (alexsherstinsky) [inline great_expectations/rule_based_profiler/toolkit.py]: @cdkini Type hint, please -- for aesthetics purposes.  Thanks! :)

File: great_expectations/rule_based_profiler/toolkit.py [hunk 13]
```diff
@@ -2,18 +2,80 @@
...
+
+    key: Union[GeCloudIdentifier, ConfigurationIdentifier]
+    if data_context.ge_cloud_mode and ge_cloud_id:
+        key = GeCloudIdentifier(resource_type="contract", ge_cloud_id=ge_cloud_id)
+    else:
+        key = ConfigurationIdentifier(
+            configuration_key=config.name,
+        )
+
+    profiler_store = data_context.profiler_store
```
Comment (alexsherstinsky) [inline great_expectations/rule_based_profiler/toolkit.py]: @cdkini For `profiler_store` too, please (type hint).

File: great_expectations/rule_based_profiler/toolkit.py (hunk 3 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/toolkit.py]: @cdkini I would be delighted to get on a call with a representative of the GE_CLOUD team (perhaps @roblim ) in order to figure out the best design for isolating the OSS from such "custom" entities.  Thanks! /cc @donaldheppner

File: great_expectations/rule_based_profiler/toolkit.py (hunk 7 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/toolkit.py:72]: @cdkini What happens if invalid?  Should the system exit, or can it go on?  If cannot go on, then exception is in order with a helpful message (ideally from one of the semantically meaningful exceptions in `great_expectations.exceptions`).  Thanks!

File: tests/conftest.py (hunk 5 above)
Comment (alexsherstinsky) [reply tests/conftest.py]: @cdkini I affixed the following in #4111 for the RBP-relevant fixtures (right above the `@fixture`) decorator:
```
# TODO: <Alex>ALEX -- come up with a more elegant way of excluding all Rule-Based Profiler tests and fixtures, unless minimum Python version requirements are satisfied.</Alex>
@pytest.mark.skipif(sys.version_info < (3, 7), reason="requires Python3.7")
```

Thanks.

File: tests/rule_based_profiler/test_toolkit.py [hunk 14]
```diff
@@ -32,6 +36,83 @@ def test_get_profiler_with_too_many_args_raises_error(
...
+    )
+
+    assert isinstance(profiler, RuleBasedProfiler)
+    assert profiler.name == profiler_config.name
+    assert mock_data_context.profiler_store.set.call_args == mock.call(
+        key=ge_cloud_profiler_key, value=profiler_config
+    )
+
+
+@mock.patch("great_expectations.data_context.data_context.DataContext")
```
Comment (alexsherstinsky) [inline tests/rule_based_profiler/test_toolkit.py:81]: @cdkini This might need some adjustment based on the merge of #4111 -- thanks.

**Review by alexsherstinsky — CHANGES_REQUESTED** [review]
@cdkini This is basically great -- just made some notes.  Thanks!

File: tests/conftest.py (hunk 5 above)
Comment (cdkini) [reply tests/conftest.py]: Placed the same warnings - this will probably change since you merged first but I'll ensure parity for now

File: great_expectations/data_context/data_context.py (hunk 2 above)
Comment (cdkini) [reply great_expectations/data_context/data_context.py]: That's fair - I'll revert to the old pattern but I think we need to have two next steps:
* Determine if we can remove Marshmallow shade and work towards implementing the true library
* Abstract this `load/dump` pattern into a Schema-level function to provide some more context.

File: great_expectations/rule_based_profiler/toolkit.py (hunk 3 above)
Comment (cdkini) [reply great_expectations/rule_based_profiler/toolkit.py]: The design here is mirroring the design spec - is the acceptance of this ticket gated on the treatment of the Cloud product?

File: great_expectations/rule_based_profiler/toolkit.py (hunk 7 above)
Comment (cdkini) [reply great_expectations/rule_based_profiler/toolkit.py:72]: this check function now returns a boolean. If False, we raise the appropriate exception.

See this snippet at the top of the toolkit:
```python
        raise ge_exceptions.InvalidConfigError(
            f'batch_data found in batch_request cannot be saved to ProfilerStore "{data_context.profiler_store_name}"'
        )
```

File: great_expectations/data_context/data_context.py (hunk 9 above)
Comment (cdkini) [reply great_expectations/data_context/data_context.py]: that's fine - I'll have to perform some merge conflicts but happy to do so

File: great_expectations/data_context/data_context.py (hunk 10 above)
Comment (cdkini) [reply great_expectations/data_context/data_context.py:3279]: I'm just working off of the spec - has this been discussed in detail? I know we're aiming to move away from this `ge_cloud_id` pattern but is this the place to draw the line?

Happy to make any necessary changes but we should discuss this live.

File: great_expectations/data_context/data_context.py (hunk 11 above)
Comment (cdkini) [reply great_expectations/data_context/data_context.py:3281]: I don't think we need type annotations for objects that are obvious; for example, `config_data` is very clearly a dictionary here.

I think type annotations make more sense in function signatures and places where there's variability but always adding them no matter what detracts from the readability of the code (in my opinion of course).

We should reflect the decision here in the style guide. cc @donaldheppner

File: great_expectations/rule_based_profiler/toolkit.py (hunk 12 above)
Comment (cdkini) [reply great_expectations/rule_based_profiler/toolkit.py]: I agree with it's placement here because `instantiate_class_from_config` returns any business logic object.

File: great_expectations/data_context/data_context.py (hunk 8 above)
Comment (cdkini) [reply great_expectations/data_context/data_context.py]: nope you're 100% correct. removed!

File: great_expectations/rule_based_profiler/toolkit.py (hunk 6 above)
Comment (cdkini) [reply great_expectations/rule_based_profiler/toolkit.py:20]: I think we're going to keep it this way for now and refactor down the line.

File: tests/rule_based_profiler/test_toolkit.py (hunk 14 above)
Comment (cdkini) [reply tests/rule_based_profiler/test_toolkit.py:81]: I think this work as is! Let me know if that's not the case.

**Review by alexsherstinsky — APPROVED** [review]
LGTM!

File: great_expectations/data_context/data_context.py [hunk 15]
```diff
@@ -3264,6 +3264,38 @@ def run_checkpoint(
...
 
+    def add_profiler(
+        self,
+        name: str,
+        config_version: float,
+        rules: Dict[str, dict],
+        class_name: str,
+        module_name: Optional[str] = None,
+        variables: Optional[dict] = None,
+        commented_map: Optional[CommentedMap] = None,
```
Comment (donaldheppner) [inline great_expectations/data_context/data_context.py]: is there a config object we could use here instead, or a business object? What's the use-case for commented map?

File: great_expectations/data_context/data_context.py [hunk 16]
```diff
@@ -3264,6 +3264,38 @@ def run_checkpoint(
             **kwargs,
         )
 
+    def add_profiler(
+        self,
+        name: str,
+        config_version: float,
+        rules: Dict[str, dict],
```
Comment (donaldheppner) [inline great_expectations/data_context/data_context.py:3275]: can this be Dict[str, Rules]? Where do the values that would be passed to this method come from?

File: great_expectations/data_context/data_context.py (hunk 11 above)
Comment (donaldheppner) [reply great_expectations/data_context/data_context.py:3281]: in this case the object's lifecycle is short lived and the IDE can fully infer the type. I'm fine omitting it in this case

File: great_expectations/data_context/data_context.py (hunk 1 above)
Comment (donaldheppner) [reply great_expectations/data_context/data_context.py]: can we take data_context out?

File: great_expectations/data_context/data_context.py (hunk 10 above)
Comment (donaldheppner) [reply great_expectations/data_context/data_context.py:3279]: we're going ahead with `ge_cloud_id` ... for now

File: great_expectations/rule_based_profiler/toolkit.py (hunk 6 above)
Comment (donaldheppner) [reply great_expectations/rule_based_profiler/toolkit.py:20]: I'd like to parameterize what we can out of data context to reduce future work, such as passing in the profiler store.

File: tests/conftest.py [hunk 17]
```diff
@@ -5083,6 +5084,20 @@ def profiler_key(profiler_name: str) -> ConfigurationIdentifier:
     return ConfigurationIdentifier(configuration_key=profiler_name)
 
 
+@pytest.fixture(scope="function")
```
Comment (donaldheppner) [inline tests/conftest.py]: `"function"` is the default scope and can be omitted

**Review by donaldheppner — CHANGES_REQUESTED** [review]

File: great_expectations/rule_based_profiler/toolkit.py [hunk 18]
```diff
@@ -14,6 +16,61 @@
...
+            f'batch_data found in batch_request cannot be saved to ProfilerStore "{data_context.profiler_store_name}"'
+        )
+
+    profiler_config: dict = config.to_json_dict()
+
+    new_profiler: RuleBasedProfiler = instantiate_class_from_config(
+        config=profiler_config,
+        runtime_environment={"data_context": data_context},
+        config_defaults={"module_name": "great_expectations.rule_based_profiler"},
+    )
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/toolkit.py]: Remove instantiate_class_from_config?

File: great_expectations/data_context/data_context.py (hunk 15 above)
Comment (cdkini) [reply great_expectations/data_context/data_context.py]: I don't believe commented_map is actually necessary here - we require it when reading from YAML but this method shouldn't need it.

File: great_expectations/rule_based_profiler/toolkit.py (hunk 6 above)
Comment (cdkini) [reply great_expectations/rule_based_profiler/toolkit.py:20]: Added in the profiler store and made a note to remove DataContext when we're in a good place to do so

File: great_expectations/data_context/data_context.py (hunk 16 above)
Comment (cdkini) [reply great_expectations/data_context/data_context.py:3275]: I don't believe so - the schema requires primitive types (see `RuleBasedProfilerConfig`'s constructor).

This method will either take the result of parsed YAML or be invoked by the user in their own script. I think either way, it's a dictionary at this point in the program.

File: great_expectations/rule_based_profiler/toolkit.py [hunk 19]
```diff
@@ -14,6 +15,65 @@
...
+    # Necessary to use instantiate_class_from_config due to the possible presence of class_name/module_name in config
+    new_profiler: RuleBasedProfiler = instantiate_class_from_config(
+        config=config.to_json_dict(),
+        runtime_environment={
+            "data_context": data_context,
+        },
+        config_defaults={
+            "module_name": "great_expectations.rule_based_profiler",
+        },
+    )
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/toolkit.py]: @donaldheppner we use `instantiate_class_from_config` here because when reading from a YAML configuration, we have `class_name` and `module_name` in the config.

We could manually pop those values from the resulting dict if we'd like (both ways don't feel too great).

Thoughts?

File: great_expectations/data_context/data_context.py [hunk 20]
```diff
@@ -3264,6 +3267,39 @@ def run_checkpoint(
...
+        config: RuleBasedProfilerConfig = RuleBasedProfilerConfig(**profiler_config)
+
+        # Chetan - 20220127 - Open to refactor all Profiler CRUD from toolkit to class methods
+        return profiler_toolkit.add_profiler(
+            config=config,
+            data_context=self,
+            profiler_store=self.profiler_store,
+            ge_cloud_id=ge_cloud_id,
+        )
+
```
Comment (alexsherstinsky) [inline great_expectations/data_context/data_context.py:3303]: @cdkini Is there a reason why prerequisite operations are not all in the `toolkit` like we have in `Checkpoint`?..

File: great_expectations/rule_based_profiler/toolkit.py [hunk 21]
```diff
@@ -14,6 +15,65 @@
...
+            batch_requests.append(domain_builder["batch_request"])
+
+        parameter_builders = rule.get("parameter_builders", [])
+        for parameter_builder in parameter_builders:
+            if "batch_request" in parameter_builder:
+                batch_requests.append(parameter_builder["batch_request"])
+
+    # DataFrames shouldn't be saved to ProfilerStore
+    for batch_request in batch_requests:
+        if batch_request_contains_batch_data(batch_request=batch_request):
```
Comment (alexsherstinsky) [inline great_expectations/rule_based_profiler/toolkit.py:72]: ❤️ ❤️ ❤️

File: great_expectations/rule_based_profiler/toolkit.py (hunk 19 above)
Comment (donaldheppner) [reply great_expectations/rule_based_profiler/toolkit.py]: I think that unless we explicitly support indirection/substitution of the `RuleBasedProfiler` type (and I don't think we do), I think we can both remove `class_name` and `module_name` from the YAML and not use `instantiate_class_from_config`.

**Review by alexsherstinsky — APPROVED** [review]
LGTM -- pending the resolution on the correct decision for `DataContext`/`toolkit` tradeoff and how to instantiate the `RuleBasedProfiler` business object.

**Review by donaldheppner — APPROVED** [review]
